File: examples/training/05-observability-guardrails/step2-input-validation.ts
import { createAgent } from '@agentuity/runtime';
import { z } from 'zod';

// Define schema for user preferences
const UserPreferencesSchema = z.object({
  language: z.enum(['en', 'es', 'fr']),
  maxResults: z.number().int().min(1).max(50)
});

// Define schema for user query
const UserQuerySchema = z.object({
  query: z.string().min(1).max(500),
  userId: z.string(),
  preferences: UserPreferencesSchema.optional()
});

type UserQuery = z.infer<typeof UserQuerySchema>;

export default createAgent('input-validator', {
  handler: async (ctx, input) => {
    // Validate input with Zod
    const parseResult = UserQuerySchema.safeParse(input);
    if (!parseResult.success) {
      ctx.logger.warn('Request validation failed', {
        errors: parseResult.error.errors
      });
      throw new Error('Invalid request: ' + parseResult.error.message);
    }

    const { query, userId, preferences } = parseResult.data;

    ctx.logger.info('Request validated successfully', {
      userId,
      queryLength: query.length,
      hasPreferences: !!preferences
    });

    // Process with confidence that data is valid
    return {
      message: 'Query processed successfully',
      query,
      userId,
      preferences: preferences || { language: 'en', maxResults: 10 },
      processedAt: new Date().toISOString()
    };
  }
});

File: examples/training/05-observability-guardrails/step2-input-validation.py
from agentuity import AgentRequest, AgentResponse, AgentContext
from pydantic import BaseModel, Field, ValidationError
from typing import Optional, Literal
from datetime import datetime


class UserPreferences(BaseModel):
    language: Literal["en", "es", "fr"]
    max_results: int = Field(ge=1, le=50)


class UserQuery(BaseModel):
    query: str = Field(min_length=1, max_length=500)
    user_id: str
    preferences: Optional[UserPreferences] = None


async def run(request: AgentRequest, response: AgentResponse, ctx: AgentContext):
    """Validates incoming requests with Pydantic schemas."""
    ctx.logger.info("Validating incoming request")

    try:
        raw_data = await request.data.json()
        # Validate with Pydantic
        validated = UserQuery(**raw_data)
    except ValidationError as e:
        ctx.logger.warning("Request validation failed", {"errors": e.errors()})
        return response.json(
            {"error": "Invalid request", "details": e.errors()}, status=400
        )

    # Use validated data (type-safe)
    ctx.logger.info(
        "Request validated successfully",
        {
            "userId": validated.user_id,
            "queryLength": len(validated.query),
            "hasPreferences": validated.preferences is not None,
        },
    )

    # Process with confidence that data is valid
    result = {
        "message": "Query processed successfully",
        "query": validated.query,
        "userId": validated.user_id,
        "preferences": (
            validated.preferences.model_dump()
            if validated.preferences
            else {"language": "en", "maxResults": 10}
        ),
        "processedAt": datetime.now().isoformat(),
    }

    return response.json(result)
